Initial working fix - Resolves date line related issues.

zach-snell opened the pull request:
Resolves issues around date line and rotation with infinite scrolling longitudinal map. May still need clean up / optimization.

JaffaKetchup reviewed (changes requested):
Hi @zach-snell, thanks for opening a PR!

I would be grateful if you could give a little more info as to what this PR fixes.

As I'm testing it, I'm comparing it to the master branch, and I can't see any difference in behaviour. Could you record some videos to show what failed before, and the fixed behaviour with your PR? I'm just struggling to understand why these changes are necessary.

As far as I understand, #2019 is completely unrelated to unbounded horizontal scrolling, and instead about issues with the `TileLayer` not being able to be changed properly. Could you explain how your PR fixes and relates to #2019?

zach-snell commented:
@JaffaKetchup 

You know, I think you're right. I was half asleep I guess and didn't relate the valid issue.

I will record a video showing the issue I was resolving later this morning. 

--

The issue was with the way the camera / tiles would flicker or jump from east to west as you move back and forth near the date line.

zach-snell commented:
Alright - adding some visuals here.
I was way down the rabbit hole through flutter_map plugins related.

I suppose the issues with flickering were at that layer but as I determined I was unable to resolve them there, there had to be a big in the core. That is what brought me to resolving this tiny yet incredibly annoying edge-case where you all luckily had a visual example already built that showed it well.

**This shows as the following in main branch**
<img src="https://github.com/user-attachments/assets/33115071-d5da-49d6-87be-074b6bd914f3" width="270" />



**Resolved, no issue with lat/long and screen smoothing**
<img src="https://github.com/user-attachments/assets/5e3816a2-7850-4096-bee0-b3bd2c5eb972" width="270" />


**For more context, downstream in flutter_vector_map_tiles**
This is after I upgraded a bunch of code to even be able to work with flutter_map 8.1.0.
<img src="https://github.com/user-attachments/assets/b1d39e1d-d8b5-488e-99e8-cba2995dbc81" width="270" />

zach-snell commented on lib/src/misc/offsets.dart:
```diff
@@ -76,22 +77,36 @@ List<Offset> getOffsetsXY({
         result = addedWidth;
-        bestX = x;
+        bestError = error;
         continue;
       }
-      if ((bestX + ox - halfScreenWidth).abs() >
-          (x + ox - halfScreenWidth).abs()) {
+
+      // Only switch worlds if there's a significant improvement
+      // This prevents oscillation near the boundary
+      if (error < bestError - hysteresisThreshold) {
         result = addedWidth;
-        bestX = x;
+        bestError = error;
       }
```
I'll have to give this a test - I had also noticed that the second assignment of bestX / bestError seemed unused. I knew I would need to review that anyways. 

Further, I remember this was related but I was iterating so many times I will confirm if reverting has impact and if so if we can simplify the logic.

zach-snell commented:
Turns out I was able to reproduce the dateline issue when I removed the drag logic. I don't know why I couldn't reproduce before. Flutter messing with me - I even ran flutter clean, flutter pub clean, deleted all build artifacts etc but could not reproduce earlier.

Attaching demonstration of that issue and the resolved version.

**Demonstrating issue**
<img src="https://github.com/user-attachments/assets/7077f928-66b0-4f7d-92e3-122b96b01a96" width="270" />

**Working resolution**
<img src="https://github.com/user-attachments/assets/b49457fd-2d9d-478f-8320-e25c98486363" width="270" />

Feel free to test on main to see if you see the same thing.
Once I reproduced, I was able to quickly change back and forth and hot-reload to see the issue.

zach-snell commented:
Alright! Very much simplified -- turns out there was a chain of self-created issues in there.
I had slowly found the resolution and part of that path introduced unnecessary code which created problems I then resolved.

This should be much easier to review!

JaffaKetchup commented:
Sorry, been focused on some other PRs. I'm still not sure what this PR is fixing. As far I can see, I can move across the dateline completely fine in the example app on that page.

zach-snell commented:
Honestly it fixes what seems to be a very small thing that had downstream
impacts.

It has to do with angles when crossing the dateline causing some of the
system to mis-represent the current position.

It has been a bit since I messed with it now but that is the jist.

The example video you can look for when the dot that stays in one spot on
the viewport starts moving in unexpected ways.


On Thu, Mar 20, 2025, 1:02 PM Luka S ***@***.***> wrote:

> Sorry, been focused on some other PRs. I'm still not sure what this PR is
> fixing. As far I can see, I can move across the dateline completely fine in
> the example app on that page.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/fleaflet/flutter_map/pull/2048#issuecomment-2741270044>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACRD2YJUAQQX4CIFQOP5MGT2VL7EZAVCNFSM6AAAAABYIXE7QOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONBRGI3TAMBUGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> [image: JaffaKetchup]*JaffaKetchup* left a comment
> (fleaflet/flutter_map#2048)
> <https://github.com/fleaflet/flutter_map/pull/2048#issuecomment-2741270044>
>
> Sorry, been focused on some other PRs. I'm still not sure what this PR is
> fixing. As far I can see, I can move across the dateline completely fine in
> the example app on that page.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/fleaflet/flutter_map/pull/2048#issuecomment-2741270044>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACRD2YJUAQQX4CIFQOP5MGT2VL7EZAVCNFSM6AAAAABYIXE7QOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONBRGI3TAMBUGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

JaffaKetchup commented:
I can't reproduce anything like is shown in the video at all. Are you using a mouse on a real device?

JaffaKetchup commented:
For the time being, I'm going to close this, because I'm still unsure if there's an issue. Whilst I can see something strange in the screen recording, I can't reproduce this. If you're still having the issue, please file a bug report!
